The cli doesn't exit correctly when stdout is a non-tty

<!-- 
THIS IS NOT A HELP FORUM. 
If you are experiencing problems with setting up Jest, please make sure to visit our Help page: 
https://facebook.github.io/jest/help.html
-->

<!-- 
Before creating an issue please check the following:
* you are using the latest version of Jest
* try re-installing your node_modules folder 
* run Jest once with `--no-cache` to see if that fixes the problem you are experiencing. 
-->

**Do you want to request a _feature_ or report a _bug_?**

Bug

**What is the current behavior?**

If you are redirecting stdout to a file (or have stdout not connected to a tty in a similar way), then the jest-cli process will exit with 0 even if there are failing tests.  If you execute jest from the cli without any kind of redirect in place, then you get a non-zero exit value for failing tests as expected.

It appears that this was introduced with the exit changes in this PR https://github.com/facebook/jest/pull/5313

**What is the expected behavior?**

A non-zero exit value in the case of test failures.  

**Please provide your exact Jest configuration and mention your Jest, node,
yarn/npm version and operating system.**

* node v6.11.4
* npm v3.10.10
* jest 22.1.1
* ubuntu 14.04

**NOTE** I updated the description since the original bug report with my latest findings

I see the problem even at the command line locally, 21.2.1 exits properly:

```
Test Suites: 1 failed, 2 passed, 3 total
Tests:       1 failed, 9 passed, 10 total
Snapshots:   0 total
Time:        2.292s
Ran all test suites.
npm ERR! Test failed.  See above for more details.

Process finished with exit code 1

```

22.0.0 is when it started failing to exit properly (and it doesn't even seem to catch its own errors, I can't show the full error, but here is what I can show):

```
      at Promise._settlePromiseFromHandler (node_modules/bluebird/js/release/promise.js:512:31)
      at Promise._settlePromise (node_modules/bluebird/js/release/promise.js:569:18)
      at Promise._settlePromise0 (node_modules/bluebird/js/release/promise.js:614:10)
      at Promise._settlePromises (node_modules/bluebird/js/release/promise.js:689:18)
      at Async.Object.<anonymous>.Async._drainQueue (node_modules/bluebird/js/release/async.js:133:16)
      at Async.Object.<anonymous>.Async._drainQueues (node_modules/bluebird/js/release/async.js:143:10)
      at Immediate.Async.drainQueues [as _onImmediate] (node_modules/bluebird/js/release/async.js:17:14)

PASS  test/mc-persistence.test.js

Test Suites: 3 passed, 3 total
Tests:       10 passed, 10 total
Snapshots:   0 total
Time:        3.555s
Ran all test suites.

Process finished with exit code 0

```

--- a/packages/jest-cli/src/cli/index.js
+++ b/packages/jest-cli/src/cli/index.js
@@ -113,7 +113,7 @@ const readResultsAndExit = (
 ) => {
   const code = !result || result.success ? 0 : globalConfig.testFailureExitCode;
 
-  process.on('exit', () => exit(code));
+  process.on('exit', () => (process.exitCode = code));
 
   if (globalConfig.forceExit) {
     exit(code);

--- a/packages/jest-runtime/src/cli/index.js
+++ b/packages/jest-runtime/src/cli/index.js
@@ -11,7 +11,6 @@ import type {Argv} from 'types/Argv';
 import type {EnvironmentClass} from 'types/Environment';
 
 import chalk from 'chalk';
-import exit from 'exit';
 import os from 'os';
 import path from 'path';
 import yargs from 'yargs';
@@ -43,7 +42,7 @@ export function run(cliArgv?: Argv, cliInfo?: Array<string>) {
 
   if (argv.help) {
     yargs.showHelp();
-    process.on('exit', () => exit(1));
+    process.on('exit', () => (process.exitCode = 1));
     return;
   }
 
@@ -54,7 +53,7 @@ export function run(cliArgv?: Argv, cliInfo?: Array<string>) {
 
   if (!argv._.length) {
     console.log('Please provide a path to a script. (See --help for details)');
-    process.on('exit', () => exit(1));
+    process.on('exit', () => (process.exitCode = 1));
     return;
   }
 
@@ -93,6 +92,6 @@ export function run(cliArgv?: Argv, cliInfo?: Array<string>) {
     })
     .catch(e => {
       console.error(chalk.red(e.stack || e));
-      process.on('exit', () => exit(1));
+      process.on('exit', () => (process.exitCode = 1));
     });
 }

